staging/src/k8s.io/apiserver/pkg/audit/context.go

```diff
@@ -46,6 +46,13 @@ type AuditContext struct {
 	annotationMutex sync.Mutex
 }
 
+// VisitEventLocked safely access AuditContext#Event map. Avoid concurrent read and write
+func (ac *AuditContext) VisitEventLocked(f func(event *auditinternal.Event)) {
```

Comment:
since the Event remains exported, and enough things grab the Event directly out of the context and mess with it (see direct accesses, calls to AuditEventFrom, WithAuditID, LogRequestMetadata, LogImpersonatedUser, LogRequestObject, LogRequestPatch, LogResponseObject, etc) that I'm not confident this getter will catch all things that access Event

Reply:
Thanks for your review 😄, I will reorganize the code based on your suggestions. I plan to make these changes:

- Change AuditContext#Event to AuditContext#event [not exported]
- Remove AuditEventFrom
- Adjust relevant codes related to event reading and writing.

Do you have any other suggestions?

Reply:
I suspect a systemic rework like that is needed to actually ensure this is threadsafe everywhere, but I also think a change that big will be too disruptive to backport... I haven't dug deeply enough to see if there's a more targeted thing we could confidently backport.

@tallclair might remember from some of the initial audit annotation mutex work (https://github.com/kubernetes/kubernetes/pull/109078) if there were next steps to make this more systemically threadsafe

Reply:
ref: https://github.com/kubernetes/kubernetes/issues/109087

Reply:
> Perhaps, we can merge the first commit (https://github.com/kubernetes/kubernetes/issues/120507 reproduced), but UT may fail in the short term. WDYT? @liggitt @tallclair

I'm glad we have a reproducer test, but we wouldn't merge a failing test without a fix... flaking/failing tests doesn't really help us

Reply:
> xref: https://github.com/kubernetes/kubernetes/issues/109087
> 
> Looks like there is a long term plan being pushed forward by @tallclair.

thanks for digging that up... I'm not sure any progress on that issue is being made currently, but the direction still sounds good to me

Reply:
I want to try following the design idea and try to reorganize the generation and reading and writing process of AuditEvent.😊

Reply:
I'll keep the current unit tests for my reproduction of #120507.  If there are related PRs that are merged later.  This unit test will verify their correctness

Reply:
thank you for your support.  When I'm done with it, I'll ping you on the current PR.

Reply:
@sxllwx Any idea if you will be picking this back up soon?

Reply:
Sorry, I haven't pushed this forward for a while.

I will complete the following tasks this week:

- [ ] Change AuditContext#Event to AuditContext#event [not exported]
- [ ] Remove AuditEventFrom
- [ ] While modifying the relevant code (core code and test code), try to clearly define the interfaces that AuditContext can provide

If possible, after I complete some of the work, I can ping you to review the related changes?